lua/wire/wireshared.lua

```diff
@@ -1212,3 +1212,56 @@ hook.Add("PlayerDisconnected", "WireLib_PlayerDisconnect", function(ply)
     tbl[ply] = nil
   end
 end)
+
+
+-- LocalToWorld, but for use in hot loops to produce as little garbage as possible
+-- by Fasteroid
+do
+	local ents_orientations      = {}
+	local ents_cached_positions = {}
+
+	local LocalToWorld = FindMetaTable("Entity").LocalToWorld
+	local orientation, cur_cached_positions, cur_ent
+
+	function WireLib.LocalToWorld_UseEnt(ent)
+		if cur_ent == ent then return end -- call this as much as you want
+
+		orientation          = ents_orientations[ent]
+		cur_cached_positions = ents_cached_positions[ent]
+		cur_ent              = ent
+
+		if not orientation or not cur_cached_positions then
+			ents_orientations[ent]     = { ent:GetPos(), ent:GetAngles() }
+			cur_cached_positions       = {}
+			ents_cached_positions[ent] = cur_cached_positions
+			return
+		end
+
+		local pos = ent:GetPos()
+		local ang = ent:GetAngles()
+
+		if not rawequal(orientation[1], pos) or not rawequal(orientation[2], ang) then
+			orientation[1]             = pos
+			orientation[2]             = ang
+			cur_cached_positions       = {}
+			ents_cached_positions[ent] = cur_cached_positions
+			return
+		end
+
+		op = retrieve
+	end
+
+	function WireLib.LocalToWorld_Find(pos)
+		local fetch = cur_cached_positions[pos]
+		if fetch then return fetch
+		else
+			cur_cached_positions[pos] = LocalToWorld(cur_ent, pos)
+			return cur_cached_positions[pos]
+		end
+	end
```

Comment:
Could this be generalized to something cleaner like
```lua
local ents_computed_data = setmetatable({}, {__index=function(t,k) local r={pos = Vector(), ang = Angle(), computed = {}} t[k]=r return r end})
function WireLib.ComputeIfEntityTransformDirty(ent, key, compute)
    local data = ents_computed_data[ent]

    local pos, ang = ent:GetPos(), ent:GetAngles()
    if not (rawequal(orientation[1], data.pos) and rawequal(orientation[2], data.ang)) then
        data.pos = pos
        data.ang = ang
        data.computed[key] = compute(ent)
    end
    return data.computed[key]
end

local wires = WireLib.ComputeIfEntityTransformDirty(ent, key, function(ent)
    return {<list of transformed vectors>}
end)
```

Reply:
Problem with my implementation is data.computed[key] will be nil if the function is first called on an entity that has position and angle at 0,0,0

Reply:
This should also prevent memory leaking caused by the vector lookup table

Reply:
Even better:

```lua
function WireLib.GetComputeIfEntityTransformDirty(compute)
    local ents_computed_data = setmetatable({}, {__index=function(t,k) local r={pos = Vector(math.huge), ang = Angle()} t[k]=r return r end})

    return function(ent)
        local data = ents_computed_data[ent]

        local pos, ang = ent:GetPos(), ent:GetAngles()
        if not (rawequal(orientation[1], data.pos) and rawequal(orientation[2], data.ang)) then
            data.pos = pos
            data.ang = ang
            data.computed = compute(ent)
        end
        return data.computed
    end
end

-- Example usage
local wireTransformer = WireLib.GetComputeIfEntityTransformDirty(function(ent)
    return {<list of transformed vectors>}
end)

local wirePositions = wireTransformer(ent)
```

Reply:
Good idea to put it on the entity itself.  Will look into doing it like that this weekend.

Reply:
I meant to do this much longer ago.  Gonna actually do it now.

Also, fuck off github-actions bot.

Reply:
> `data.computed = compute(ent, ...)`
Varargs not jittable, going to avoid this